### ISSUE TITLE
Cannot run multi-GPU inference on another model in the same process after YOLO

### ISSUE THREAD

**Issue by middleknight** [ultralytics/ultralytics#24853]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I found that this issue is caused by ultralytics/utils/torch_utils.py. 

`os.environ["CUDA_VISIBLE_DEVICES"] = device  # set environment variable - must be before assert is_available()`

Why is this operation necessary? If I need to perform multi-GPU inference after this YOLO inference, how should I do it? 

My current workaround is to revert this change after the inference. Will this cause any side effects?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @middleknight, thank you for your interest in Ultralytics 🚀! This is an automated response to help get things moving, and an Ultralytics engineer will also assist you soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
Since this appears to involve runtime behavior around device handling and multi-GPU inference, please share a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) showing:
-Your Ultralytics version
-Your PyTorch version
-Your CUDA version
-Your OS
-A minimal script that reproduces the issue
-What model you run before and after YOLO
-How you launch inference and what device string you pass
-Any full traceback or unexpected behavior you observe
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
-**Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
-**Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
-**Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
-**Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You can initialize CUDA before inference with

```
import torch

torch.cuda.init()
```

Then the environment variable update will not have an effect.

And for multi-GPU, you should use a context manager:

https://github.com/ultralytics/ultralytics/issues/17973#issuecomment-2516828801

**Comment by middleknight** [comment]
I tried adding torch.cuda.init() before inference, but it had no effect. os.environ["CUDA_VISIBLE_DEVICES"] is still overwritten after the first inference.

For the multi-GPU part, my goal is to run separate models. For instance: GPU 0 -> YOLO, GPU 1 -> SAM3, GPU 2 -> others. Under this loading sequence, if I do a warmup for YOLO, the later models cannot find the remaining GPUs.

**Comment by Y-T-G** [comment]
> I tried adding torch.cuda.init() before inference, but it had no effect. os.environ["CUDA_VISIBLE_DEVICES"] is still overwritten after the first inference.

`CUDA_VISIBLE_DEVICES` would be overwritten but after initialization, it will have no effect, so it doesn't matter. All the GPUs would be visible regardless of `CUDA_VISIBLE_DEVICES` because CUDA was already initialized before. `CUDA_VISIBLE_DEVICES` only has an effect if you initialize CUDA after setting it. It has no effect if CUDA is initialized before it's updated.

> For the multi-GPU part, my goal is to run separate models. For instance: GPU 0 -> YOLO, GPU 1 -> SAM3, GPU 2 -> others. Under this loading sequence, if I do a warmup for YOLO, the later models cannot find the remaining GPUs.

This shouldn't occur if you initialized CUDA before running inference and used the `torch.cuda.device()` context manager I linked above.

**Comment by middleknight** [comment]
Problem solved. Thanks a lot!

**Comment by glenn-jocher** [comment]
Glad to hear it’s resolved, thanks for confirming.
